cudax/include/cuda/experimental/__launch/launch_transform.cuh

```diff
@@ -0,0 +1,81 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX__LAUNCH_LAUNCH_TRANSFORM
+#define _CUDAX__LAUNCH_LAUNCH_TRANSFORM
+#include <cuda/__cccl_config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__cuda/api_wrapper.h>
+#include <cuda/std/__type_traits/decay.h>
+#include <cuda/std/utility>
+#include <cuda/stream_ref>
+
+#include <cuda/experimental/__detail/utility.cuh>
+
+#if _CCCL_STD_VER >= 2017
+namespace cuda::experimental::detail
+{
+// Types should define overloads of __cudax_launch_transform that are find-able
+// by ADL in order to customize how cudax::launch handles that type. The
+// overload below, which simply returns the argument unmodified, is the overload
+// that gets chosen if no other overload matches. It takes __ignore as the first
+// argument to make this overload less preferred than other overloads that take
+// a stream_ref as the first argument.
+template <typename _Arg>
+_Arg&& __cudax_launch_transform(__ignore, _Arg&& __arg) noexcept
+{
+  return _CUDA_VSTD::forward<_Arg>(__arg);
+}
+
+template <typename _Arg>
+using __launch_transform_direct_result_t =
+  decltype(__cudax_launch_transform(::cuda::stream_ref{}, _CUDA_VSTD::declval<_Arg>()));
+
+struct __fn
+{
+  template <typename _Arg>
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE __launch_transform_direct_result_t<_Arg>
+  operator()(::cuda::stream_ref __stream, _Arg&& __arg) const
+  {
+    // This call is unqualified to allow ADL
+    return __cudax_launch_transform(__stream, _CUDA_VSTD::forward<_Arg>(__arg));
+  }
+};
+
+template <typename _Arg, typename _Enable = void>
+struct __launch_transform_result
+{
+  using type = _CUDA_VSTD::decay_t<__launch_transform_direct_result_t<_Arg>>;
+};
+
+template <typename _Arg>
+struct __launch_transform_result<
+  _Arg,
+  _CUDA_VSTD::void_t<typename _CUDA_VSTD::decay_t<__launch_transform_direct_result_t<_Arg>>::__launch_transform_result>>
+{
+  using type = typename _CUDA_VSTD::decay_t<__launch_transform_direct_result_t<_Arg>>::__launch_transform_result;
+};
+
+template <typename _Arg>
+using __launch_transform_result_t = typename __launch_transform_result<_Arg>::type;
```

Comment:
I think this alias needs to be in the main namespace, so the user can apply it when instantiating kernel templates.
Similar to how `cudax::finalized_t` is a public alias in #2001, might even make sense to unify it one both get merged

Reply:
Ok, I can see this mechanism to not be available for the users, but I still think we need the public alias to be able to do a template specialization in generic context

Reply:
> I still think we need the public alias to be able to do a template instantiation in generic context

I don't understand.

Reply:
Lets say you have a kernel in a library like this:
```
template <typename T>
__global__ void lib_kernel_foo(T input) {...}
```

If you are a library writer and expect some range of inputs for lib_kernel_foo and want to use cudax::launch. You will do something like this:
```
template <typename T>
void do_foo(const T& input) {
    //...
    auto kernel = lib_kernel_foo<T>;
    cudax::launch(conf, kernel, input);
}
```
But in cases where launch transforms T it will fail. They would need to do something like this:
```
auto kernel = lib_kernel_foo<cudax::transformed_t<T>>;
```

Reply:
I see. Yes that makes sense, thanks.